Появляется ошибка

Lidia128 opened the pull request:
Error starting ApplicationContext. To display the conditions report re-run your application with 'debug' enabled.
                  2023-01-12 23:16:54.490 ERROR 9500 --- [  restartedMain] o.s.boot.SpringApplication               : Application run failed

Lidia128 committed:
Появляется ошибка
Error starting ApplicationContext. To display the conditions report re-run your application with 'debug' enabled.
                  2023-01-12 23:16:54.490 ERROR 9500 --- [  restartedMain] o.s.boot.SpringApplication               : Application run failed

OlgaLugacheva commented on src/main/java/me/recipe/homework/service/impl/RecipeServiceImpl.java:
```diff
@@ -3,42 +3,62 @@
-        this.filesRecipeService = filesRecipeService;
+    private final Path pathToTxtTemplate;
+
+    public RecipeServiceImpl(@Value("./recipeData.json") FilesRecipeService filesRecipeService) {
```
Value нужно отсюда убрать, сервис - это не строка, поэтому эту аннотацию применить здесь нельзя

OlgaLugacheva commented on src/main/java/me/recipe/homework/service/impl/RecipeServiceImpl.java:
```diff
@@ -3,42 +3,62 @@
     private static int id = 0;
-    public RecipeServiceImpl(FilesRecipeService filesRecipeService) {
-        this.filesRecipeService = filesRecipeService;
+    private final Path pathToTxtTemplate;
```
вот это в этом классе быть не должно, сервис рецептов не должен значть ничего о файлах , когда у нас появляется ответвенный за это сервис раьботы с файлами

OlgaLugacheva commented on src/main/java/me/recipe/homework/service/impl/RecipeServiceImpl.java:
```diff
@@ -3,42 +3,62 @@
+    public RecipeServiceImpl(@Value("./recipeData.json") FilesRecipeService filesRecipeService) {
+        try {
+            this.filesRecipeService = filesRecipeService;
+            this.pathToTxtTemplate = Paths.get(RecipeServiceImpl.class.getResource("recipeTemplate.txt").toURI());
```
это лишнее

OlgaLugacheva commented on src/main/java/me/recipe/homework/service/impl/RecipeServiceImpl.java:
```diff
@@ -91,6 +118,34 @@ public Path createRecipesFile() {
+    @Override
+    public byte[] exportTxt() {
+        try {
+            String template = Files.readString(pathToTxtTemplate, StandardCharsets.UTF_8);
```
вот это - это задача FileService но не этого класса

Lidia128 committed:
Ошибки не ушли

OlgaLugacheva commented on src/main/java/me/recipe/homework/service/impl/FilesRecipeServiceImpl.java:
```diff
@@ -1,21 +1,48 @@
+    final private FilesRecipeService filesRecipeService;
+    static Map<Integer, Recipe> recipes = new TreeMap<>();
+    private static int id = 0;
+    private final Path pathToTxtTemplate;
```
я бы вообще удалила это поле

OlgaLugacheva commented on src/main/java/me/recipe/homework/service/impl/FilesRecipeServiceImpl.java:
```diff
@@ -1,21 +1,48 @@
 
 @Service
 public class FilesRecipeServiceImpl implements FilesRecipeService {
+    final private FilesRecipeService filesRecipeService;
```
вот это поле убрать

OlgaLugacheva commented on src/main/java/me/recipe/homework/service/impl/FilesRecipeServiceImpl.java:
```diff
@@ -1,21 +1,48 @@
     @Value("${name.of.recipe.data.fale}")
     private String dataFileName;
 
+    public FilesRecipeServiceImpl(FilesRecipeService filesRecipeService) {
+        this.filesRecipeService = filesRecipeService;
+        try {
+            this.pathToTxtTemplate = Paths.get(RecipeServiceImpl.class.getResource("recipeTemplate.txt").toURI());
+        } catch (InvalidPathException e) {
+            e.printStackTrace();
+            throw e;
+        } catch (URISyntaxException e) {
```
уберите вообще этот конструктор, здесь у вас и происходит циклическая зависимость
получение пути из строки можно написать в методе где этот .pathToTxtTemplate используется аналогично другим путям

OlgaLugacheva commented on src/main/java/me/recipe/homework/service/impl/RecipeServiceImpl.java:
```diff
@@ -3,43 +3,48 @@
-    public RecipeServiceImpl(FilesRecipeService filesRecipeService) {
-        this.filesRecipeService = filesRecipeService;
+
+    public RecipeServiceImpl(RecipeService recipeService) {
     }
```
убрать

OlgaLugacheva commented on src/main/java/me/recipe/homework/service/impl/RecipeServiceImpl.java:
```diff
@@ -91,16 +103,21 @@ public Path createRecipesFile() {
     }
     private void readFromFile() {
-        String json = filesRecipeService.readFromFile();
+        String json = RecipeService.readFromFile();
```
вот этот метод почему-то всегда вовзарщает null - какая тут должна быть логика ?

OlgaLugacheva commented on src/main/java/me/recipe/homework/service/impl/RecipeServiceImpl.java:
```diff
@@ -91,16 +103,21 @@ public Path createRecipesFile() {
-        String json = filesRecipeService.readFromFile();
+        String json = RecipeService.readFromFile();
         try {
             recipes = new ObjectMapper().readValue(json, new TypeReference<TreeMap<Integer, Recipe>>() {
```
здесь приложение будет падать так как json всегда null

OlgaLugacheva commented on src/main/java/me/recipe/homework/service/RecipeService.java:
```diff
@@ -1,15 +1,19 @@
+    static void saveToFile(String json) {
+    }
+
+    static String readFromFile() {
+        return null;
+    }
```
почему тут null ? 
во-первых реализация должна быть не в интерфейсе, а во-вторых в ней должна быть логика чтения

OlgaLugacheva commented on src/main/java/me/recipe/homework/service/impl/RecipeServiceImpl.java:
```diff
@@ -3,43 +3,48 @@
     }
+
     @Override
     public Recipe getRecipe(int id) {
         return null;
     }
```
почему? где логика метода?

OlgaLugacheva commented on src/main/java/me/recipe/homework/service/impl/RecipeServiceImpl.java:
```diff
@@ -3,43 +3,48 @@
     }
+
     @Override
     public Recipe addRecipe(Recipe recipe) {
         return null;
```
где логика метода?

OlgaLugacheva commented on src/main/java/me/recipe/homework/service/impl/RecipeServiceImpl.java:
```diff
@@ -91,16 +103,21 @@ public Path createRecipesFile() {
         return null;
```
почему везде все методы вовзращают null ?

OlgaLugacheva commented on src/main/java/me/recipe/homework/service/impl/RecipeServiceImpl.java:
```diff
@@ -48,17 +53,20 @@ public Recipe addNewRecipe(Recipe recipe) {
+
     @Override
     public Path createRicepe(Recipe recipe) throws IOException {
         return null;
```
почему?

OlgaLugacheva commented on src/main/java/me/recipe/homework/service/impl/RecipeServiceImpl.java:
```diff
@@ -67,21 +75,25 @@ public Recipe editRecipe(long id, Recipe recipe) {
         }
         return null;
```
это неверно

OlgaLugacheva commented on src/main/java/me/recipe/homework/service/impl/RecipeServiceImpl.java:
```diff
@@ -67,21 +75,25 @@ public Recipe editRecipe(long id, Recipe recipe) {
+
     @Override
     public Recipe updateRecipe(int id, Recipe recipe) {
         return null;
```
это неверно

Lidia128 committed:
Попыталась внести корректировки, но вышло не очень

OlgaLugacheva commented on src/main/java/me/recipe/homework/service/impl/RecipeServiceImpl.java:
```diff
@@ -3,43 +3,44 @@
     }
+
     @Override
     public Recipe addRecipe(Recipe recipe) {
```
в этом методе нужно положить элемент в коллекцию 
recipes.put(id++, recipe)
но коллекция должна принадлежать этому классу а не быть импортом из FileREcipeServiceImpl

OlgaLugacheva commented on src/main/java/me/recipe/homework/service/impl/RecipeServiceImpl.java:
```diff
@@ -3,43 +3,44 @@
+
     @Override
     public Recipe getRecipe(int id) {
-        return null;
```
нужно обраиться к коллекции и вернуть по id рецепт из нее

OlgaLugacheva commented on src/main/java/me/recipe/homework/service/impl/RecipeServiceImpl.java:
```diff
@@ -48,59 +49,70 @@ public Recipe addNewRecipe(Recipe recipe) {
     @Override
     public Path createRicepe(Recipe recipe) throws IOException {
-        return null;
+        return recipe;
```
в чем суть этого метода здесь ? и почему он не path тогда возвращает

OlgaLugacheva commented on src/main/java/me/recipe/homework/service/impl/RecipeServiceImpl.java:
```diff
@@ -48,59 +49,70 @@ public Recipe addNewRecipe(Recipe recipe) {
     }
+
     @Override
     public Recipe updateRecipe(int id, Recipe recipe) {
```
уже есть editRecipe зачем тогда нужен этот метод тем более что он написан неверно ?

OlgaLugacheva commented on src/main/java/me/recipe/homework/service/impl/RecipeServiceImpl.java:
```diff
@@ -48,59 +49,70 @@ public Recipe addNewRecipe(Recipe recipe) {
     }
+
     @Override
     public Recipe removeRecipe(int id) {
```
зачем этот метод если есть уже deleteRecipe ?

OlgaLugacheva commented on src/main/java/me/recipe/homework/service/impl/FilesRecipeServiceImpl.java:
```diff
@@ -62,4 +73,33 @@ public boolean cleanDateFile() {
+    @Override
+    public byte[] exportTxt() {
+        try {
+            String template = Files.readString(pathToTxtTemplate, StandardCharsets.UTF_8);
```
почему бы вместо Path.of(dataFilePath, dataFileName)
не использовать что-то вроде Path.of(dataFilePath, dataFileName)
тем более что это используется в другом методе ?

Lidia128 committed:
Попыталась внести корректировки, но вышло не очень

Lidia128 committed:
Скорректировала код, но в FileRecipeServiseImpl ст.84 recipe.values, recipe красного цвета.

OlgaLugacheva commented on src/main/java/me/recipe/homework/service/impl/FilesRecipeServiceImpl.java:
```diff
@@ -62,4 +76,34 @@ public boolean cleanDateFile() {
+        try {
+            String template = Files.readString(Path.of(dataFilePath, dataFileName), StandardCharsets.UTF_8);
+            StringBuilder stringBuilder = new StringBuilder();
+            for (Recipe recipe : recipe.values()) {
```
вероятно, recipe это карта с рецептами, которая сейчас у вас хранится в RecipeSericeImpl, тут вопрос где вы хотите использовать метод exportTxt ? Может быть есть тогда смысл перенести эту логику туда  в тот сервис?

Lidia128 committed:
1.Перенесла метод TXT в RecipeSeviceImpl и несколько сопутствующих ему методов, но ситуация не изменилась в ст.131 recipe.values, recipe красного цвета.
2. метод saveToFile задваивается в RecipeService. Если удаляю, отражаетс проблема этого метода в RecipeServiceImpl.

OlgaLugacheva commented on src/main/java/me/recipe/homework/service/impl/RecipeServiceImpl.java:
```diff
@@ -85,22 +93,74 @@ public boolean deleteRecipe(long id) {
+        try {
+            String template = Files.readString(Path.of(dataFilePath, dataFileName), StandardCharsets.UTF_8);
+            StringBuilder stringBuilder = new StringBuilder();
+            for (Recipe recipe : recipe.values()) {
```
у вас же коллекция называет тут recipes

OlgaLugacheva commented on src/main/java/me/recipe/homework/service/impl/RecipeServiceImpl.java:
```diff
@@ -85,22 +93,74 @@ public boolean deleteRecipe(long id) {
+                                .replace("%steps%", steps.toString()));
+                stringBuilder.append(recipeData).append("\n\n\n");
+            }
+            return stringBuilder.toString().getBytes(StandardCharsets.UTF_8);
+            return template.getBytes(StandardCharsets.UTF_8);
```
почему 2 return, определиите какой оставите

OlgaLugacheva commented on src/main/java/me/recipe/homework/service/RecipeService.java:
```diff
@@ -28,8 +27,18 @@ public interface RecipeService {
+
+    File getDataFile();
+
+    boolean cleanDateFile();
```
удалить

Lidia128 committed:
1. saveToFile задваивается в RecipeService. Если удаляю, отражается проблема этого метода в RecipeServiceImpl.
2. в FileRecipeServiceImpl в ст. 38 и ст. 52 все data переменные горят красны. Я так думаю, это потому-что переменные перенесены в RecipeServiceImpl их сделать public?

OlgaLugacheva commented on src/main/java/me/recipe/homework/service/impl/RecipeServiceImpl.java:
```diff
@@ -91,16 +103,65 @@ public Path createRecipesFile() {
         try {
             String json = new ObjectMapper().writeValueAsString(recipes);
-            filesRecipeService.saveToFile(json);
+            RecipeService.saveToFile(json);
```
this.saveToFile(json);

OlgaLugacheva commented on src/main/java/me/recipe/homework/service/impl/FilesRecipeServiceImpl.java:
```diff
@@ -1,30 +1,35 @@
+import java.nio.charset.StandardCharsets;
 import java.nio.file.Files;
+import java.nio.file.InvalidPathException;
 import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.Map;
+import java.util.TreeMap;
 
 @Service
 public class FilesRecipeServiceImpl implements FilesRecipeService {
-    @Value("${path.to.recipe.data.file}")
-    private String dataFilePath;
-    @Value("${name.of.recipe.data.fale}")
-    private String dataFileName;
 
```
так как вы удалили эти поля, то теперь логично, что они неопределены, верните поля, если вы их используете

Lidia128 committed:
data переменные горят красным в зависимости от того где они находятся в RecipeServiceImpl если они завдены в FileRecipeServiceImpl и наоборот. Мне необходимо их сделать public, чтоб они везде читались или все методоы с их спользованием перенести в одно место?

Lidia128 committed:
Появляется ошибка. Копии сделала в описании к заданию.

OlgaLugacheva commented on src/main/java/me/recipe/homework/service/impl/RecipeServiceImpl.java:
```diff
@@ -3,85 +3,98 @@
     private static int id = 0;
-    public RecipeServiceImpl(FilesRecipeService filesRecipeService) {
+    private Map<Integer, Recipe> recipes;
+    private final Path pathToTextTemplate;
```
ошибка у вас из-за того, что вы не проинициализировали это поле, но я не вижу, что оно используется, возможно его вам нужно удалить тогда, если не используете.

OlgaLugacheva commented on src/main/java/me/recipe/homework/service/impl/RecipeServiceImpl.java:
```diff
@@ -3,85 +3,98 @@
+
+    public RecipeServiceImpl(FilesRecipeService filesRecipeService, Path pathToTextTemplate) {
         this.filesRecipeService = filesRecipeService;
+        this.pathToTextTemplate = pathToTextTemplate;
```
тоже лишнее

Lidia128 committed:
После того как я внесла изменения согласно замечаний у меня подчеркнуло всю ст. 21 и recips везед стал красным

OlgaLugacheva commented on src/main/java/me/recipe/homework/service/impl/RecipeServiceImpl.java:
```diff
@@ -23,14 +19,9 @@
 
     private final FilesRecipeService filesRecipeService;
-    private static int id = 0;
-    private Map<Integer, Recipe> recipes;
```
разве я писала, что нужно удалить счетчик, карту и конструктор? Я говорила только про поле
Верните выше обозначенное обратно

Lidia128 committed:
Скорректировала код, программа запустилась.

Lidia128 committed:
Скорректировала код, программа запустилась.
